### ISSUE TITLE
Better conversion to/from SymPy/SymEngine (or others)

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#25715]
Following on from discussion in gh-25401

CC @isuruf @certik 

Currently both SymPy and SymEngine have sympify functions that can convert each others expressions:
```python
In [9]: import sympy

In [10]: import symengine

In [11]: e = symengine.cos(1)

In [12]: type(e)
Out[12]: symengine.lib.symengine_wrapper.cos

In [13]: sympy.sympify(e)
Out[13]: cos(1)

In [14]: type(sympy.sympify(e))
Out[14]: cos

In [15]: type(sympy.sympify(e)).__module__
Out[15]: 'sympy.functions.elementary.trigonometric'

In [16]: type(symengine.sympify(sympy.cos(1)))
Out[16]: symengine.lib.symengine_wrapper.cos
```
Otherwise though I am not sure what conversion functions there exist to convert between SymPy and SymEngine.

The implicit simplification is useful because it means that you can pass SymEngine expressions into a SymPy function:
```python
In [17]: x = symengine.symbols('x')

In [18]: sympy.solve(symengine.cos(x) - 1, x)
Out[18]: [0, 2⋅π]
```
However the results are returned as SymPy expressions:
```python
In [20]: type(_[0]).__module__
Out[20]: 'sympy.core.numbers'
```
It would be useful to be able to pass a SymEngine expression into a SymPy function and then get a SymEngine expression returned back. The problem is that the `sympify` mechanism only works in one direction.

There could be a decorator for handling this:
```python
@to_sympy
def solve(...):
    # SymEngine expressions converted to SymPy on input
    # and back to SymEngine on output
    ...
```
This would make it easier for someone to use SymPy functions while doing a calculation with SymEngine. Obviously there are lots of details that would need to be worked out to make the `to_sympy` decorator work. I envisage something like:
```python
convert_back, expr_sympy = expr._expression_converter(format='sympy')

# do calculations

return convert_back(result)
```
If we had a systematic way of performing conversions like this then it would also make it easier to use SymEngine in a localised way within SymPy itself because we could use the same mechanism for using SymEngine inside a particular function:
```python
@use_symengine
def diff(expr, x):
    # If SymEngine is installed the inputs are now converted to SymEngine
    # If the inputs were SymPy expressions they will be converted back
    ...
```
Other projects like Sage might also use this conversion mechanism for calling SymPy functions.

**Comment by asmeurer** [comment]
Wouldn't we want sympy.sympify() to *not* convert SymEngine types into SymPy types? Otherwise, it's impossible to pass a SymEngine type to a SymPy function that calls sympify() at the top (most of them), and have it actually do the computation using SymEngine. If SymEngine provided the same interface as SymPy, and if this conversion didn't happen, then doing this would just work. In principle, you could write something like 

```py
sympy.simplify(symengine_expr)
```

and it would perform the algorithm inside of sympy.simplify, but using SymEngine expressions.

**Comment by oscarbenjamin** [comment]
> Wouldn't we want sympy.sympify() to _not_ convert SymEngine types into SymPy types?

No, if the code was not written with the intention that it should work with SymEngine expressions internally then it should convert to SymPy expressions. Very few things would "just work" if sympify was changed to pass SymEngine expressions straight into the internals of random parts of SymPy. That needs to be done in a controlled way which is why I am proposing that there should be controlled conversions.

**Comment by oscarbenjamin** [comment]
I suppose one problem with this is that any expression with the old assumptions would not be able to round trip:
```python
In [4]: x = sympy.symbols('x', positive=True)

In [5]: x == sympy.sympify(symengine.sympify(x))
Out[5]: False
```
The decorator could potentially handle that but it would add to the conversion costs.

Or perhaps the `__expression_converter` could handle it by returning something that knows how to map the symbols back.

**Comment by isuruf** [comment]
It should be possible to implement Symbols with old assumptions in the python layer without polluting the C++ codebase. We didn't implement old assumptions in the C++ library, because it was uncertain what SymPy might end up with.

**Comment by oscarbenjamin** [comment]
One thing that just occurred to me:

There is one place in SymPy where conversions are already carefully controlled which is in the poly domains. For the most part the poly domains do not use symbolic expressions but actually SymEngine would provide a much faster version of the EX domain for arithmetic. Currently SymEngine expressions do not have `cancel` which is the basic simplification used in `EX` but at least within the context of the `EX` domain it would be possible to control that by converting to SymPy for `cancel` when needed.

**Comment by oscarbenjamin** [comment]
@isuruf do you have any thoughts on what could make a good conversion interface? Ideally it would be something extensible.

Currently `_sympy_` is extensible but it only works in one direction.

**Comment by isuruf** [comment]
@oscarbenjamin, how do you envision using methods like `cancel` in such a system? Do you expect the types to have a `.cancel` method and etc and not a `cancel` method where you pass the objects to?

**Comment by oscarbenjamin** [comment]
In the EX domain the domain elements are just Expr wrapped up in another class:
```python
In [1]: p = Poly(x + cos(t) + sin(t), x)

In [2]: p
Out[2]: Poly(x + sin(t) + cos(t), x, domain='EX')

In [3]: p.rep
Out[3]: DMP([EX(1), EX(sin(t) + cos(t))], EX, None)

In [4]: p.rep.rep
Out[4]: [EX(1), EX(sin(t) + cos(t))]

In [5]: type(p.rep.rep[1])
Out[5]: sympy.polys.domains.expressiondomain.ExpressionDomain.Expression

In [6]: p.rep.rep[1].ex
Out[6]: sin(t) + cos(t)

In [7]: type(p.rep.rep[1].ex)
Out[7]: sympy.core.add.Add
```
The other class wraps all the methods:
https://github.com/sympy/sympy/blob/a95eb0860343ae245499cb1ada0573495cafacdc/sympy/polys/domains/expressiondomain.py#L49-L50
Usually in the domain system if you want to make alternative implementations of a domain you would either make a separate class for the elements or just a separate class for the domain `EX` object itself. A subclass like `Expression_SymEngine` could define the methods differently and so if SymEngine does not have cancel then it could call whatever methods are available.

Ultimately cancel needs to compute the polynomial gcd and so for that part if would need to convert to SymPy I think. However a lot of the work involved in cancel is often just expand. In a division free algorithm we don't need the cancel so much: it might be possible to shortcut doing any conversions if we can see that there is no Add in the denominator.

**Comment by oscarbenjamin** [comment]
An example where division free algorithms are used is for charpoly. You can see some discussion about this with timings here: https://github.com/sympy/sympy/pull/25458#issuecomment-1662397968

**Comment by oscarbenjamin** [comment]
You can read more about the domain system and how to use it here:
https://docs.sympy.org/latest/modules/polys/domainsintro.html
